### PR TITLE
dev-lang/nim: enable docs and parallel build

### PR DESCRIPTION
**Pull request by CyberTailor**

### CODE REVIEW

File: dev-lang/nim/nim-1.6.10-r1.ebuild [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+		dev-libs/libffi
+		dev-libs/libpcre:3
+		dev-libs/openssl
+		media-libs/libsdl
+		media-libs/libsfml
+	)
+"
+
+BDEPEND="
+	sys-process/parallel
```
Comment (xgqt) [inline dev-lang/nim/nim-1.6.10-r1.ebuild:38]: So parallel works now? Does it still spam the build log with ads?

File: dev-lang/nim/nim-1.6.10-r1.ebuild [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+	mkdir "${HOME}"/.parallel || die
+	touch "${HOME}"/.parallel/will-cite || die "parallel setup failed"
+}
+
+src_compile() {
+	local -x PATH="${S}/bin:${PATH}"
+
+	edo ./build.sh --parallel "$(makeopts_jobs)"
+
+	ebegin "Waiting for unfinished jobs"
```
Comment (xgqt) [inline dev-lang/nim/nim-1.6.10-r1.ebuild:104]: No other way to do this?
Parallelism issues do not disappear if you just `sleep`...

**Comment by xgqt** [comment]
Have you run the node & valgrind tests?
Maybe we should just disable them altogether?

File: dev-lang/nim/nim-1.6.10-r1.ebuild (hunk 1 above)
Comment (CyberTailor) [reply dev-lang/nim/nim-1.6.10-r1.ebuild:38]: I touch `will-cite` file so it doesn't spam the logs

**Comment by CyberTailor** [comment]
Node tests work on my machines, valgrind don't (because i have `-march`)

**Comment by xgqt** [comment]
> Node tests work on my machine, valgrind don't (because i have `-march`)

Then you do not know if valgrind test work, no need to enable them then.

**Comment by CyberTailor** [comment]
OK, disabled them

File: dev-lang/nim/nim-1.6.10-r1.ebuild [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+	experimental? (
+		https://git.sr.ht/~cyber/${PN}-patches/archive/${PATCH_PV}.tar.gz
+			-> nim-patches-${PATCH_PV}.tar.gz
+	)
+"
+
+LICENSE="MIT"
+SLOT="0"
+KEYWORDS="~amd64 ~arm ~x86"
+IUSE="debug doc experimental +test-js test"
```
Comment (xgqt) [inline dev-lang/nim/nim-1.6.10-r1.ebuild:22]: I do not think `+test-js` has to be on by default, I think the case is similar to `test-full` on some other packages, in fact maybe we should change this here too?

File: dev-lang/nim/nim-1.6.10-r1.ebuild [hunk 4]
```diff
@@ -0,0 +1,223 @@
+# Copyright 1999-2022 Gentoo Authors
+# Distributed under the terms of the GNU General Public License v2
+
+EAPI=8
+
+inherit bash-completion-r1 edo multiprocessing toolchain-funcs xdg-utils
+
+PATCH_PV="1.6.6_p1"
```
Comment (xgqt) [inline dev-lang/nim/nim-1.6.10-r1.ebuild:8]: Style nitpick: put this above `inherit`s.

File: dev-lang/nim/nim-1.6.10-r1.ebuild (hunk 3 above)
Comment (CyberTailor) [reply dev-lang/nim/nim-1.6.10-r1.ebuild]: Done

File: dev-lang/nim/nim-1.6.10-r1.ebuild (hunk 4 above)
Comment (CyberTailor) [reply dev-lang/nim/nim-1.6.10-r1.ebuild]: Done

**Comment by xgqt** [comment]
Thanks, I am testing this on my machine right now.

**Comment by xgqt** [comment]
`tstats.nim` fails for me

```
FAIL: tests/stdlib/tstats.nim c                                    ( 2.43 sec)
Test "tests/stdlib/tstats.nim" in category "stdlib"
Failure: reExitcodesDiffer
Expected:
exitcode: 0

Gotten:
exitcode: 1

Output:
fatal.nim(54)            sysFatal
Error: unhandled exception: tstats.nim(39, 13) `rr`gensym49.intercept() == -0.03424657534246611`  [AssertionDefect]


diff --git a/var/tmp/portage/dev-lang/nim-1.6.10-r1/temp/diffStrings_a_D1wPTH83 b/var/tmp/portage/dev-lang/nim-1.6.10-r1/temp/diffStrings_b_XKCZ7AKW
index 031aaaa..0a12513 100644
--- a/var/tmp/portage/dev-lang/nim-1.6.10-r1/temp/diffStrings_a_D1wPTH83
+++ b/var/tmp/portage/dev-lang/nim-1.6.10-r1/temp/diffStrings_b_XKCZ7AKW
@@ -1 +1,5 @@
-exitcode: 0
\ No newline at end of file
+exitcode: 1
+
+Output:
+fatal.nim(54)            sysFatal
+Error: unhandled exception: tstats.nim(39, 13) `rr`gensym49.intercept() == -0.03424657534246611`  [AssertionDefect]
```

**Comment by xgqt** [comment]
> cannot open file: jester

Afaik jester is a 3rd-party library,
this one probably needs to be disabled.

```
FAIL: tests/niminaction/Chapter7/Tweeter/src/tweeter.nim c         ( 3.66 sec)
Test "tests/niminaction/Chapter7/Tweeter/src/tweeter.nim" in category "niminaction"
Failure: reNimcCrash
$ /var/tmp/portage/dev-lang/nim-1.6.10-r1/work/nim-1.6.10/bin/nim c --hints:on -d:testing --nimblePath:build/deps/pkgs   --nimCache:nimcache/tests/niminaction/Chapter7/Tweeter/src/tweeter.nim_4a8a08f09d37b73795649038408b5f33  tests/niminaction/Chapter7/Tweeter/src/tweeter.nim
Hint: used config file '/var/tmp/portage/dev-lang/nim-1.6.10-r1/work/nim-1.6.10/config/nim.cfg' [Conf]
Hint: used config file '/var/tmp/portage/dev-lang/nim-1.6.10-r1/work/nim-1.6.10/config/config.nims' [Conf]
Hint: used config file '/var/tmp/portage/dev-lang/nim-1.6.10-r1/work/nim-1.6.10/nim.cfg' [Conf]
Hint: used config file '/var/tmp/portage/dev-lang/nim-1.6.10-r1/work/nim-1.6.10/tests/config.nims' [Conf]
tweeter.nim(7, 8) Error: cannot open file: jester
```

**Comment by xgqt** [comment]
Other failing tests:

```
FAIL: tests/lent/tbasic_lent_check.nim c                           ( 9.24 sec)
Test "tests/lent/tbasic_lent_check.nim" in category "lent"
Failure: reExitcodesDiffer
Expected:
exitcode: 0

Gotten:
exitcode: 1

Output:
1
SIGSEGV: Illegal storage access. (Attempt to read from nil?)


diff --git a/var/tmp/portage/dev-lang/nim-1.6.10-r1/temp/diffStrings_a_hkmwCR9E b/var/tmp/portage/dev-lang/nim-1.6.1
0-r1/temp/diffStrings_b_t5LpYUyI
index 031aaaa..63e1175 100644
--- a/var/tmp/portage/dev-lang/nim-1.6.10-r1/temp/diffStrings_a_hkmwCR9E
+++ b/var/tmp/portage/dev-lang/nim-1.6.10-r1/temp/diffStrings_b_t5LpYUyI
@@ -1 +1,5 @@
-exitcode: 0
\ No newline at end of file
+exitcode: 1
+
+Output:
+1
+SIGSEGV: Illegal storage access. (Attempt to read from nil?)
```

**Comment by xgqt** [comment]
Full build log:

[build.log](https://github.com/gentoo/gentoo/files/10099489/build.log)

**Comment by CyberTailor** [comment]
> `tstats.nim` fails for me

What is your arch and flags?

> Full build log

If you need to share build logs, install `app-text/ansifilter` and add this to make.conf:
`PORTAGE_LOG_FILTER_FILE_CMD=ansifilter`

**Comment by CyberTailor** [comment]
Deselected tests that don't work for you

**Comment by xgqt** [comment]
there's also this weird one failing for me just now (possibly parallelism issue)

```
PASS: tests/js/tmangle.nim js                                      ( 0.50 sec)
oserr.nim(95)            raiseOSError
Error: unhandled exception: Bad file descriptor
Additional info: Could not find command: 'tests/ic/tgenerics_temp'. OS error: Text file busy [OSError]
PASS: tests/ic/tgenerics_temp.nim c  --incremental:on -d:nimIcIntegrityChecks ( 4.26 sec)
```

**Comment by CyberTailor** [comment]
ok, deselected it too

**Comment by xgqt** [comment]
And I got yet another random failure.
Were you able to reproduce any of those?

```
FAIL: compiler/nim.nim c                                           ( 7.33 sec)
Test "compiler/nim.nim" in category "ic"
Failure: reNimcCrash
$ /var/tmp/portage/dev-lang/nim-1.6.10-r1/work/nim-1.6.10/bin/nim check --hint:Conf:off --warnings:off --ic:on   --nimCache:nimcache/compiler/nim.nim_4a8a08f09d37b73795649038408b5f33  compiler/nim.nim
fatal.nim(54)            sysFatal
Error: unhandled exception: index 257 not in 0 .. 256 [IndexDefect]
```

**Comment by xgqt** [comment]
Weird, `compiler/nim.nim` is listed in the tests to ignore...

**Comment by CyberTailor** [comment]
> in category "ic"

Something from that category calls it.

**Comment by CyberTailor** [comment]
> Were you able to reproduce any of those?

No x_x

**Comment by CyberTailor** [comment]
I'll try to enable tests explicitly instead of calling "testament all"

**Comment by xgqt** [comment]
> I'll try to enable tests explicitly instead of calling "testament all"

That's a good idea!

**Comment by xgqt** [comment]
@CyberTailor 

> I'll try to enable tests explicitly instead of calling "testament all"

Any progress on this?

Maybe postpone enabling tests?

**Comment by CyberTailor** [comment]
I'm too busy with surviving right now. Might continue later

**Comment by CyberTailor** [comment]
> Maybe postpone enabling tests?

Dropped tests from this PR, with intention to enable them later.

**Comment by xgqt** [comment]
@CyberTailor you will need to bump revisions after my last commit.

Is this ready to go?

**Comment by CyberTailor** [comment]
Missed that! Updated now

**Comment by xgqt** [comment]
`IUSE=debug` is leftover there

**Comment by CyberTailor** [comment]
oops, fixed

**Comment by xgqt** [comment]
I'm testing this rn.

**Comment by xgqt** [comment]
Please, squash this into one commit with either changes description or at least listing the other commits.

**Comment by CyberTailor** [comment]
Done

**Comment by xgqt** [comment]
Amazing! Thanks a lot!
Merging it now.

**Pull request closed without merging**